plugins/jabra-file/fu-jabra-file-device.c

```diff
@@ -0,0 +1,855 @@
+			    rx_data.rxbuf[5],
+			    rx_data.rxbuf[6]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_write_next_block(FuJabraFileDevice *self,
+				      guint32 chunk_number,
+				      const guint8 *buf,
+				      gsize bufsz,
+				      GError **error)
+{
+	guint8 write_length = (((chunk_number + 1) % 101 == 0) ? (0x80) : (0x00)) + bufsz + 8;
+	FuJabraFileTxData tx_data = {
+	    .txbuf =
+		{
+		    FU_JABRA_FILE_IFACE,
+		    0x01,
+		    0x00,
+		    self->sequence_number,
+		    write_length,
+		    0x03,
+		    0x00,
+		},
+	    .timeout = FU_JABRA_FILE_STANDARD_SEND_TIMEOUT,
+	};
+	FuJabraFileRxData rx_data = {
+	    .rxbuf = {0x00},
+	    .timeout = FU_JABRA_FILE_STANDARD_RECEIVE_TIMEOUT,
+	};
+
+	guint8 data[2];
+	data[0] = (guint8)(FU_JABRA_FILE_NEXT_BLOCK << 6 | (bufsz + 1));
+	data[1] = (guint8)chunk_number;
+	if (!fu_memcpy_safe(tx_data.txbuf, FU_JABRA_FILE_BUF_SIZE, 7, data, 2, 0x0, 2, error))
+		return FALSE;
+	if (!fu_memcpy_safe(tx_data.txbuf,
+			    FU_JABRA_FILE_BUF_SIZE,
+			    9,
+			    buf,
+			    bufsz,
+			    0x0,
+			    bufsz,
+			    error))
+		return FALSE;
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_file_device_tx_cb,
+				  FU_JABRA_FILE_MAX_RETRIES,
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&tx_data,
+				  error))
+		return FALSE;
+	if ((chunk_number + 1) % 101 == 0) {
+		if (!fu_device_retry_full(FU_DEVICE(self),
+					  fu_jabra_file_device_rx_cb,
+					  FU_JABRA_FILE_MAX_RETRIES,
+					  FU_JABRA_FILE_RETRY_DELAY,
+					  (gpointer)&rx_data,
+					  error))
+			return FALSE;
+
+		if (rx_data.rxbuf[5] != 0xFF) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "internal error: expected 0xFF, got 0x%02x 0x%02x",
+				    rx_data.rxbuf[5],
+				    rx_data.rxbuf[6]);
+			return FALSE;
+		}
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_write_blocks(FuJabraFileDevice *self,
+				  FuChunkArray *chunks,
+				  FuProgress *progress,
+				  GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+	for (gint chunk_number = 0; (guint)chunk_number < fu_chunk_array_length(chunks);
+	     chunk_number++) {
+		g_autoptr(FuChunk) chk = NULL;
+
+		/* prepare chunk */
+		chk = fu_chunk_array_index(chunks, chunk_number, error);
+		if (chk == NULL)
+			return FALSE;
+		if (!fu_jabra_file_device_write_next_block(self,
+							   chunk_number,
+							   fu_chunk_get_data(chk),
+							   fu_chunk_get_data_sz(chk),
+							   error))
+			return FALSE;
+		fu_progress_step_done(progress);
+	}
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_check_device_busy(FuJabraFileDevice *self, GError **error)
+{
+	FuJabraFileTxData tx_data = {
+	    .txbuf =
+		{
+		    FU_JABRA_FILE_IFACE,
+		    self->address,
+		    0x00,
+		    self->sequence_number,
+		    0x46,
+		    0x26,
+		    0x1D,
```

Comment:
are these numbers special? if so, maybe hash define them?

Reply:
could they be 0x0?

Reply:
these are values for the communication protocol towards the device. so they could be anything, but of course we want them to be the right values to achieve the right messages back and forth